File: system/src/main/scala/model/Status.scala
package lila.system
package model

sealed abstract class Status(val id: Int) extends Ordered[Status] {

  def compare(other: Status) = id compare other.id
}

case object Created extends Status(10)
case object Started extends Status(20)
case object Aborted extends Status(25)
case object Mate extends Status(30)
case object Resign extends Status(31)
case object Stalemate extends Status(32)
case object Timeout extends Status(33)
case object Draw extends Status(34)
case object Outoftime extends Status(35)
case object Cheat extends Status(36)

object Status {

  val all = List(Created, Started, Aborted, Mate, Resign, Stalemate, Timeout, Draw, Outoftime, Cheat)

  val byId = all map { v ⇒ (v.id, v) } toMap

  def apply(id: Int): Option[Status] = byId get id
}

File: ui/lib/src/game/status.ts
import type { GameData } from './interfaces';

// https://github.com/lichess-org/scalachess/blob/master/core/src/main/scala/Status.scala

export const status: { [name in StatusName]: number } = {
  created: 10,
  started: 20,
  aborted: 25,
  mate: 30,
  resign: 31,
  stalemate: 32,
  timeout: 33,
  draw: 34,
  outoftime: 35,
  cheat: 36,
  noStart: 37,
  unknownFinish: 38,
  insufficientMaterialClaim: 39,
  variantEnd: 60,
};

export type StatusName =
  | 'created'
  | 'started'
  | 'aborted'
  | 'mate'
  | 'resign'
  | 'stalemate'
  | 'timeout'
  | 'draw'
  | 'insufficientMaterialClaim'
  | 'outoftime'
  | 'noStart'
  | 'cheat'
  | 'variantEnd'
  | 'unknownFinish';

export interface Status {
  id: StatusId;
  name: StatusName;
}

export type StatusId = number;

export const statusOf = (name: StatusName): Status => ({ id: status[name], name });

export const started = (data: GameData): boolean => data.game.status.id >= status.started;

export const finished = (data: GameData): boolean => data.game.status.id >= status.mate;

export const aborted = (data: GameData): boolean => data.game.status.id === status.aborted;

export const playing = (data: GameData): boolean => started(data) && !finished(data) && !aborted(data);
